substrate/client/network/src/request_responses.rs

```diff
@@ -594,6 +618,59 @@ impl NetworkBehaviour for RequestResponsesBehaviour {
 
 	fn poll(&mut self, cx: &mut Context) -> Poll<ToSwarm<Self::ToSwarm, THandlerInEvent<Self>>> {
 		'poll_all: loop {
+			// Poll the periodic request check.
+			if let Poll::Ready(_) = self.periodic_request_check.poll_tick(cx) {
+				// First pass is needed to collect timed out requests and
+				// submit `OutboundFailure::Timeout` on the moved oneshot sender.
+				let timedout_requests: Vec<_> = self
+					.pending_requests
+					.iter()
+					.filter_map(|(id, req)| {
```

Comment:
Why not retain? Then we don't need collect & remove.

Reply:
There's a tiny borrow-checker issue coming from the sender consuming self: retain will give us `&mut req` and the `req.response` oneshot::Sender consumes `self` when we send out the `send(Err(RequestFailure::Network(OutboundFailure::Timeout)))`.

We could still avoid the extra alloc by adding an Option over the sender, will have a look if it complicates the code, thanks